File: eventfinder/scraper.py
"""The scraping pipeline.

Rather than screen-scraping one site's fragile HTML, we pull the structured
schema.org ``Event`` data (JSON-LD) that most event sites already embed for
search engines. That makes the same extractor work across Meetup, Eventbrite,
venue and city-tourism pages — you just point it at more source URLs.

Pipeline:  fetch source pages -> extract Event JSON-LD -> geocode each venue
-> compute distance from the user -> drop anything past the range -> tag.
"""
from __future__ import annotations

import html
import json
import unicodedata
from datetime import datetime, timedelta, timezone
from typing import Iterable, Optional
from urllib.parse import quote

import requests
from bs4 import BeautifulSoup

from . import geo, tagging
from .places import fetch_places
from .models import Event, event_id

BROWSER_UA = (
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) "
    "Chrome/125.0 Safari/537.36"
)


# --------------------------------------------------------------------------
# JSON-LD extraction
# --------------------------------------------------------------------------
def _iter_ld_nodes(data) -> Iterable[dict]:
    """Yield every dict node in a JSON-LD payload, unwrapping @graph/lists."""
    stack = [data]
    while stack:
        node = stack.pop()
        if isinstance(node, list):
            stack.extend(node)
        elif isinstance(node, dict):
            if "@graph" in node and isinstance(node["@graph"], list):
                stack.extend(node["@graph"])
            yield node


def _is_event(node: dict) -> bool:
    t = node.get("@type", "")
    types = t if isinstance(t, list) else [t]
    return any("Event" in str(x) for x in types)


def _first(value):
    return value[0] if isinstance(value, list) and value else value


def _text(value) -> str:
    value = _first(value)
    if isinstance(value, dict):
        value = value.get("name") or value.get("url") or ""
    if not isinstance(value, str):
        return ""
    return html.unescape(value).strip()


def _location_fields(node: dict) -> tuple[str, str, Optional[float], Optional[float]]:
    """Return (address, locality, lat, lon) from an Event's location, if physical."""
    loc = _first(node.get("location"))
    if not isinstance(loc, dict):
        return "", "", None, None
    if "Virtual" in str(loc.get("@type", "")):
        return "", "", None, None  # online-only event, skip

    lat = lon = None
    geo_node = loc.get("geo")
    if isinstance(geo_node, dict):
        try:
            lat = float(geo_node["latitude"])
            lon = float(geo_node["longitude"])
        except (KeyError, ValueError, TypeError):
            lat = lon = None

    addr = loc.get("address")
    locality = ""
    parts: list[str] = []
    if isinstance(addr, dict):
        locality = addr.get("addressLocality", "") or ""
        for key in ("streetAddress", "addressLocality", "postalCode", "addressCountry"):
            val = addr.get(key)
            if val and val not in parts:
                parts.append(str(val))
    elif isinstance(addr, str):
        parts.append(addr)

    name = loc.get("name", "")
    address = ", ".join(p for p in ([name] if name else []) + parts if p)
    return html.unescape(address), html.unescape(locality), lat, lon


def extract_events(html: str, source_name: str) -> list[Event]:
    soup = BeautifulSoup(html, "lxml")
    events: list[Event] = []
    seen: set[str] = set()
    for tag in soup.find_all("script", attrs={"type": "application/ld+json"}):
        raw = tag.string or tag.get_text() or ""
        raw = raw.strip()
        if not raw:
            continue
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            continue
        for node in _iter_ld_nodes(data):
            if not _is_event(node):
                continue
            title = _text(node.get("name"))
            url = _text(node.get("url"))
            start = _text(node.get("startDate"))
            if not title:
                continue
            eid = event_id(url, title, start)
            if eid in seen:
                continue
            seen.add(eid)
            address, locality, lat, lon = _location_fields(node)
            description = _text(node.get("description"))
            events.append(
                Event(
                    id=eid,
                    title=title,
                    url=url,
                    source=source_name,
                    description=description,
                    start=start,
                    end=_text(node.get("endDate")),
                    image=_text(node.get("image")),
                    address=address,
                    locality=locality,
                    lat=lat,
                    lon=lon,
                    auto_tags=tagging.auto_tags(title, description),
                )
            )
    return events


# --------------------------------------------------------------------------
# Source fetching
# --------------------------------------------------------------------------
def build_source_url(template: str, place: dict, city_lower_override: Optional[str] = None) -> str:
    city = (place.get("city") or place.get("label", "").split(",")[0]).strip()
    # allevents-style slugs want a lowercased, hyphenated city.
    city_slug = quote(city_lower_override) if city_lower_override else quote(city.lower().replace(" ", "-"))
    return template.format(
        city=quote(city),
        city_lower=city_slug,
        cc=quote((place.get("cc") or "").strip()),
        label=quote(place.get("label", "")),
    )


def _strip_diacritics(s: str) -> str:
    return "".join(
        c for c in unicodedata.normalize("NFKD", s) if not unicodedata.combining(c)
    )


def city_slug_candidates(place: dict) -> list[str]:
    """Ordered slug guesses for a city — event sites localize names badly
    (Nominatim gives "Hanover", but AllEvents wants "hannover"; "München" must
    become "munich", not "muenchen"). We try the local name, an ASCII-folded
    version, and the English name, and let the caller pick by proximity."""
    names = [n for n in (place.get("city_local"), place.get("city")) if n]
    if not names and place.get("label"):
        names = [place["label"].split(",")[0]]
    out: list[str] = []
    for name in names:
        base = name.strip().lower().replace(" ", "-")
        for slug in (base, _strip_diacritics(base)):
            if slug and slug not in out:
                out.append(slug)
    return out


def resolve_allevents_slug(
    place: dict, origin_lat: float, origin_lon: float, range_km: float
) -> tuple[Optional[str], Optional[str]]:
    """Pick the AllEvents city slug whose /all events are actually near the
    origin. Returns (slug, cached_all_html) so the winning page can be reused."""
    first: tuple[Optional[str], Optional[str]] = (None, None)
    for slug in city_slug_candidates(place):
        html = fetch(f"https://allevents.in/{quote(slug)}/all")
        if first[0] is None:
            first = (slug, html)
        if not html:
            continue
        events = extract_events(html, "AllEvents")
        if any(
            e.lat is not None
            and geo.haversine_km(origin_lat, origin_lon, e.lat, e.lon) <= range_km
            for e in events
        ):
            return slug, html
    return first


def fetch(url: str) -> Optional[str]:
    try:
        resp = requests.get(
            url,
            headers={"User-Agent": BROWSER_UA, "Accept-Language": "en,de"},
            timeout=25,
        )
        if resp.status_code == 200:
            return resp.text
    except requests.RequestException:
        return None
    return None


# --------------------------------------------------------------------------
# Full run
# --------------------------------------------------------------------------
def _apply_manual(ev: Event, overrides: dict) -> None:
    """Combine auto tags with the user's manual add/remove for this event."""
    tags = set(ev.auto_tags)
    entry = overrides.get(ev.id)
    if entry:
        tags.update(entry.get("add", []))
        tags.difference_update(entry.get("remove", []))
    ev.tags = sorted(tags)


def _parse_start(start: str) -> Optional[datetime]:
    """Parse an ISO-8601 start string to an aware datetime, or None."""
    if not start:
        return None
    s = start.strip().replace("Z", "+00:00")
    try:
        dt = datetime.fromisoformat(s)
    except ValueError:
        # try just the date part, e.g. "2026-07-22"
        try:
            dt = datetime.fromisoformat(s[:10])
        except ValueError:
            return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt


def _within_window(dt: Optional[datetime], now: datetime, days: int) -> Optional[bool]:
    """True if within the window, False if outside, None if undated."""
    if dt is None:
        return None
    return (now - timedelta(days=1)) <= dt <= (now + timedelta(days=days))


def run_search(
    place: dict,
    sources: list[dict],
    range_km: float,
    store,
    log: Optional[list] = None,
    window_days: int = 365,
    include_places: bool = True,
) -> list[Event]:
    """Find things to do within ``range_km``: scheduled events happening in the
    next ``window_days`` days (plus undated/recurring listings), and permanent
    places you can visit any day (museums, parks, …)."""
    log = log if log is not None else []
    geocode_cache = store.load_geocode_cache()
    manual = store.load_manual_tags()
    origin_lat, origin_lon = place["lat"], place["lon"]
    now = datetime.now(timezone.utc)

    # Work out which AllEvents city slug actually returns local events (city
    # names localize inconsistently), reusing the winning /all page.
    enabled = [s for s in sources if s.get("enabled", True)]
    has_allevents = any("allevents.in" in s.get("url", "") for s in enabled)
    ae_slug: Optional[str] = None
    ae_all_html: Optional[str] = None
    if has_allevents:
        ae_slug, ae_all_html = resolve_allevents_slug(place, origin_lat, origin_lon, range_km)
        if ae_slug:
            log.append(f"AllEvents city slug: {ae_slug}")

    # Fetch sources sequentially: AllEvents returns a degraded, IP-geolocated
    # response when hit with concurrent requests, so we must not parallelize.
    collected: dict[str, Event] = {}
    for src in enabled:
        is_allevents = "allevents.in" in src.get("url", "")
        # Reuse the /all page we already fetched while picking the slug.
        if is_allevents and src["url"].endswith("/all") and ae_all_html is not None:
            html = ae_all_html
        else:
            url = build_source_url(
                src["url"], place, city_lower_override=ae_slug if is_allevents else None
            )
            html = fetch(url)
        if html is None:
            log.append(f"{src['name']}: could not fetch")
            continue
        found = extract_events(html, src["name"])
        log.append(f"{src['name']}: {len(found)} event(s) with structured data")
        for ev in found:
            collected.setdefault(ev.id, ev)

    results: list[Event] = []
    dropped_old = 0
    for ev in collected.values():
        # Only keep events happening within the next week (undated listings,
        # which are often recurring, are kept and shown without a date).
        in_window = _within_window(_parse_start(ev.start), now, window_days)
        if in_window is False:
            dropped_old += 1
            continue
        # Geocode the venue if the feed didn't already carry coordinates.
        if ev.lat is None or ev.lon is None:
            query = ev.address or ev.locality
            if not query:
                continue
            loc = geo.geocode(query, cache=geocode_cache)
            if loc is None:
                continue
            ev.lat, ev.lon = loc.lat, loc.lon
        ev.distance_km = round(
            geo.haversine_km(origin_lat, origin_lon, ev.lat, ev.lon), 1
        )
        if ev.distance_km > range_km:
            continue
        ev.direction = (
            geo.compass16(origin_lat, origin_lon, ev.lat, ev.lon)
            if ev.distance_km > 0 else ""
        )
        _apply_manual(ev, manual)
        results.append(ev)
    store.save_geocode_cache(geocode_cache)
    if dropped_old:
        log.append(f"skipped {dropped_old} event(s) outside the next {window_days} days")

    # Permanent places (museums, parks, galleries, …) you can visit any day.
    if include_places:
        places = fetch_places(origin_lat, origin_lon, range_km)
        added = 0
        for ev in places:
            if ev.id in {e.id for e in results}:
                continue
            ev.distance_km = round(
                geo.haversine_km(origin_lat, origin_lon, ev.lat, ev.lon), 1
            )
            if ev.distance_km > range_km:
                continue
            ev.direction = (
                geo.compass16(origin_lat, origin_lon, ev.lat, ev.lon)
                if ev.distance_km > 0 else ""
            )
            _apply_manual(ev, manual)
            results.append(ev)
            added += 1
        log.append(f"{added} permanent place(s) from OpenStreetMap")

    # Upcoming dated events first (by date), then undated events, then permanent
    # places (nearest first).
    def sort_key(e: Event):
        if e.permanent:
            return (2, e.distance_km or 0.0, "")
        dt = _parse_start(e.start)
        if dt is None:
            return (1, e.distance_km or 0.0, "")
        return (0, dt.timestamp(), "")

    results.sort(key=sort_key)
    log.append(f"{len(results)} thing(s) to do within {range_km:g} km")
    return results

File: android/app/src/main/java/com/therealdeltrex/localeventfinder/data/Scraper.kt
package com.therealdeltrex.localeventfinder.data

import android.text.Html
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import org.jsoup.Jsoup
import java.time.Instant
import java.time.LocalDate
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.time.temporal.ChronoUnit

/**
 * The scraping pipeline, ported from the web app's scraper.py.
 *
 * Reads the structured schema.org `Event` data (JSON-LD) that most event sites
 * embed for search engines, so one extractor works across many sites. Offline
 * events with a geocodable address are kept; online-only events are skipped.
 */
class Scraper(private val client: OkHttpClient, private val geo: Geo) {

    private val places = Places(client)

    private val browserUa =
        "Mozilla/5.0 (Linux; Android 13) AppleWebKit/537.36 (KHTML, like Gecko) " +
            "Chrome/125.0 Mobile Safari/537.36"

    // ---- fetching --------------------------------------------------------
    private suspend fun fetch(url: String): String? = withContext(Dispatchers.IO) {
        val req = Request.Builder()
            .url(url)
            .header("User-Agent", browserUa)
            .header("Accept-Language", "en,de")
            .build()
        try {
            client.newCall(req).execute().use { resp ->
                if (resp.code == 200) resp.body?.string() else null
            }
        } catch (e: Exception) {
            null
        }
    }

    // ---- AllEvents city-slug resolution ---------------------------------
    private fun stripDiacritics(s: String): String =
        java.text.Normalizer.normalize(s, java.text.Normalizer.Form.NFKD)
            .replace(Regex("\\p{M}+"), "")

    private fun nameToSlugs(name: String): List<String> {
        val base = name.trim().lowercase().replace(" ", "-")
        val out = LinkedHashSet<String>()
        if (base.isNotEmpty()) out.add(base)
        val ascii = stripDiacritics(base)
        if (ascii.isNotEmpty()) out.add(ascii)
        return out.toList()
    }

    private data class LocalityAll(val slug: String?, val html: String?, val near: Int)

    /** Try slug variants of a place name against AllEvents /all. */
    private suspend fun fetchAllEventsAll(
        name: String, olat: Double, olon: Double, rangeKm: Double,
    ): LocalityAll {
        var best = LocalityAll(null, null, 0)
        for (slug in nameToSlugs(name)) {
            val html = fetch("https://allevents.in/${java.net.URLEncoder.encode(slug, "UTF-8")}/all")
                ?: continue
            val near = extractEvents(html, "AllEvents").count {
                it.lat != null && Geo.haversineKm(olat, olon, it.lat!!, it.lon!!) <= rangeKm
            }
            if (near > 0) return LocalityAll(slug, html, near)
            if (best.html == null) best = LocalityAll(slug, html, near)
        }
        return best
    }

    /** Collect AllEvents listings for the origin city AND nearby towns within
     *  the radius, so a remote location still finds events nearby. */
    private suspend fun gatherAllEvents(
        place: Place, olat: Double, olon: Double, rangeKm: Double, log: MutableList<String>,
    ): List<Event> {
        val names = LinkedHashSet<String>()
        for (n in listOf(place.cityLocal, place.city)) if (n.isNotBlank()) names.add(n)
        // A town centre can sit just outside the range while its events fall in.
        for (n in places.nearbyLocalities(olat, olon, rangeKm + 20)) names.add(n)

        val out = ArrayList<Event>()
        val seen = HashSet<String>()
        val scored = ArrayList<Triple<Int, String, String>>()  // near, slug, name
        fun add(events: List<Event>) { for (e in events) if (seen.add(e.id)) out.add(e) }

        for (name in names.take(8)) {
            val r = fetchAllEventsAll(name, olat, olon, rangeKm)
            if (r.html == null) continue
            add(extractEvents(r.html, "AllEvents ($name)"))
            if (r.near > 0 && r.slug != null) {
                scored.add(Triple(r.near, r.slug, name))
                log.add("AllEvents · $name: ${r.near} nearby event(s)")
            }
        }
        scored.sortByDescending { it.first }
        scored.take(1).forEach { (_, slug, name) ->
            for (cat in listOf("music", "theatre", "comedy", "arts", "nightlife")) {
                val html = fetch("https://allevents.in/${java.net.URLEncoder.encode(slug, "UTF-8")}/$cat")
                if (html != null) add(extractEvents(html, "AllEvents ($name/$cat)"))
            }
        }
        return out
    }

    // ---- JSON-LD extraction ---------------------------------------------
    /** Collect every JSON object node, unwrapping @graph and arrays. */
    private fun collectNodes(root: Any, out: MutableList<JSONObject>) {
        when (root) {
            is JSONArray -> for (i in 0 until root.length()) collectNodes(root.get(i), out)
            is JSONObject -> {
                out.add(root)
                val graph = root.opt("@graph")
                if (graph is JSONArray) collectNodes(graph, out)
            }
        }
    }

    private fun isEvent(node: JSONObject): Boolean {
        val t = node.opt("@type") ?: return false
        val types = if (t is JSONArray) (0 until t.length()).map { t.getString(it) } else listOf(t.toString())
        return types.any { it.contains("Event") }
    }

    private fun text(value: Any?): String {
        val v = when (value) {
            is JSONArray -> if (value.length() > 0) value.get(0) else ""
            else -> value
        }
        val s = when (v) {
            is JSONObject -> v.optString("name", v.optString("url", ""))
            null -> ""
            else -> v.toString()
        }
        return Html.fromHtml(s, Html.FROM_HTML_MODE_LEGACY).toString().trim()
    }

    private data class Loc(
        val address: String, val locality: String, val lat: Double?, val lon: Double?,
    )

    private fun locationFields(node: JSONObject): Loc {
        val raw = node.opt("location")
        val loc = when (raw) {
            is JSONArray -> if (raw.length() > 0) raw.optJSONObject(0) else null
            is JSONObject -> raw
            else -> null
        } ?: return Loc("", "", null, null)

        if (loc.optString("@type", "").contains("Virtual")) return Loc("", "", null, null)

        var lat: Double? = null
        var lon: Double? = null
        loc.optJSONObject("geo")?.let { g ->
            lat = g.optString("latitude").toDoubleOrNull()
            lon = g.optString("longitude").toDoubleOrNull()
        }

        var locality = ""
        val parts = ArrayList<String>()
        when (val addr = loc.opt("address")) {
            is JSONObject -> {
                locality = addr.optString("addressLocality", "")
                for (key in listOf("streetAddress", "addressLocality", "postalCode", "addressCountry")) {
                    val v = addr.optString(key, "")
                    if (v.isNotEmpty() && v !in parts) parts.add(v)
                }
            }
            is String -> if (addr.isNotEmpty()) parts.add(addr)
        }
        val name = loc.optString("name", "")
        val address = (if (name.isNotEmpty()) listOf(name) else emptyList()).plus(parts)
            .filter { it.isNotEmpty() }.joinToString(", ")
        return Loc(unescape(address), unescape(locality), lat, lon)
    }

    private fun unescape(s: String) =
        Html.fromHtml(s, Html.FROM_HTML_MODE_LEGACY).toString()

    fun extractEvents(html: String, sourceName: String): List<Event> {
        val doc = Jsoup.parse(html)
        val events = ArrayList<Event>()
        val seen = HashSet<String>()
        for (script in doc.select("script[type=application/ld+json]")) {
            val raw = script.data().trim()
            if (raw.isEmpty()) continue
            val parsed: Any = runCatching { JSONObject(raw) }.getOrNull()
                ?: runCatching { JSONArray(raw) }.getOrNull()
                ?: continue
            val nodes = ArrayList<JSONObject>()
            collectNodes(parsed, nodes)
            for (node in nodes) {
                if (!isEvent(node)) continue
                val title = text(node.opt("name"))
                if (title.isEmpty()) continue
                val url = text(node.opt("url"))
                val start = text(node.opt("startDate"))
                val eid = eventId(url, title, start)
                if (!seen.add(eid)) continue
                val loc = locationFields(node)
                val description = text(node.opt("description"))
                events.add(
                    Event(
                        id = eid,
                        title = title,
                        url = url,
                        source = sourceName,
                        description = description,
                        start = start,
                        end = text(node.opt("endDate")),
                        image = text(node.opt("image")),
                        address = loc.address,
                        locality = loc.locality,
                        lat = loc.lat,
                        lon = loc.lon,
                        autoTags = Tagging.autoTags(title, description),
                    )
                )
            }
        }
        return events
    }

    // ---- full run --------------------------------------------------------
    /**
     * Scrape enabled sources, geocode, distance-filter to [rangeKm], and apply
     * the user's manual tag overrides. [log] receives human-readable progress.
     */
    suspend fun runSearch(
        place: Place,
        sources: List<Source>,
        rangeKm: Double,
        overrides: Map<String, TagOverride>,
        log: MutableList<String>,
        windowDays: Long = 365,
        includePlaces: Boolean = true,
    ): List<Event> {
        val enabled = sources.filter { it.enabled }
        val collected = LinkedHashMap<String, Event>()

        // Non-AllEvents sources (e.g. Meetup) — queried by the origin city.
        for (src in enabled) {
            if (src.url.contains("allevents.in")) continue
            val html = fetch(Sources.buildUrl(src.url, place))
            if (html == null) {
                log.add("${src.name}: could not fetch")
                continue
            }
            val found = extractEvents(html, src.name)
            log.add("${src.name}: ${found.size} event(s) with structured data")
            for (ev in found) collected.putIfAbsent(ev.id, ev)
        }

        // AllEvents: origin city AND nearby towns within the radius.
        if (enabled.any { it.url.contains("allevents.in") }) {
            for (ev in gatherAllEvents(place, place.lat, place.lon, rangeKm, log)) {
                collected.putIfAbsent(ev.id, ev)
            }
        }

        val now = Instant.now()
        val windowStart = now.minus(1, ChronoUnit.DAYS)
        val windowEnd = now.plus(windowDays, ChronoUnit.DAYS)
        val results = ArrayList<Event>()
        var droppedOld = 0
        for (ev in collected.values) {
            // Keep events in the next `windowDays`; undated listings are kept.
            val dt = parseStart(ev.start)
            if (dt != null && (dt.isBefore(windowStart) || dt.isAfter(windowEnd))) {
                droppedOld++
                continue
            }
            if (ev.lat == null || ev.lon == null) {
                val query = ev.address.ifEmpty { ev.locality }
                if (query.isEmpty()) continue
                val loc = geo.geocode(query) ?: continue
                ev.lat = loc.lat
                ev.lon = loc.lon
            }
            val d = Geo.haversineKm(place.lat, place.lon, ev.lat!!, ev.lon!!)
            ev.distanceKm = Math.round(d * 10) / 10.0
            if (ev.distanceKm!! > rangeKm) continue
            ev.direction = if (ev.distanceKm!! > 0) Geo.compass16(place.lat, place.lon, ev.lat!!, ev.lon!!) else ""
            ev.tags = effectiveTags(ev, overrides[ev.id])
            results.add(ev)
        }
        if (droppedOld > 0) log.add("skipped $droppedOld event(s) outside the next $windowDays days")

        // Permanent places you can visit any day (OpenStreetMap).
        if (includePlaces) {
            val existing = results.map { it.id }.toHashSet()
            var added = 0
            for (ev in places.fetch(place.lat, place.lon, rangeKm)) {
                if (ev.id in existing) continue
                val d = Geo.haversineKm(place.lat, place.lon, ev.lat!!, ev.lon!!)
                ev.distanceKm = Math.round(d * 10) / 10.0
                if (ev.distanceKm!! > rangeKm) continue
                ev.direction = if (ev.distanceKm!! > 0) Geo.compass16(place.lat, place.lon, ev.lat!!, ev.lon!!) else ""
                ev.tags = effectiveTags(ev, overrides[ev.id])
                results.add(ev)
                added++
            }
            log.add("$added permanent place(s) from OpenStreetMap")
        }

        // Upcoming dated events first (by date), then undated events, then
        // permanent places (nearest first).
        results.sortWith(compareBy({ sortGroup(it) }, { sortValue(it) }))
        log.add("${results.size} thing(s) to do within ${rangeKm.toInt()} km")
        return results
    }

    private fun parseStart(s: String): Instant? {
        if (s.isBlank()) return null
        return try {
            OffsetDateTime.parse(s).toInstant()
        } catch (e: Exception) {
            try {
                LocalDate.parse(s.take(10)).atStartOfDay(ZoneOffset.UTC).toInstant()
            } catch (e2: Exception) {
                null
            }
        }
    }

    private fun sortGroup(e: Event): Int =
        if (e.permanent) 2 else if (parseStart(e.start) == null) 1 else 0

    private fun sortValue(e: Event): Double {
        val dt = if (e.permanent) null else parseStart(e.start)
        return dt?.toEpochMilli()?.toDouble() ?: (e.distanceKm ?: 0.0)
    }

    private fun effectiveTags(ev: Event, override: TagOverride?): Set<String> {
        val tags = ev.autoTags.toMutableSet()
        if (override != null) {
            tags.addAll(override.add)
            tags.removeAll(override.remove)
        }
        return tags
    }
}
